tests/xfstests/run.pm

```diff
@@ -433,10 +437,81 @@ sub config_debug_option {
     }
 }
 
+# Run a single test and write log to file but without heartbeat
+sub test_run_without_hb {
+    my ($self, $test) = @_;
+    my ($category, $num) = split(/\//, $test);
+    my $run_options = '';
+    my $status_num = 1;
+    $run_options = '-nfs' if check_var('XFSTESTS', 'nfs');
+    my $inject_code = get_var('INJECT_INFO', '');
+    eval {
+        $test_start = time();
+        # Send kill signal 3 seconds after sending the default SIGTERM to avoid some tests refuse to stop after timeout
+        assert_script_run("timeout -k 3 " . ($TIMEOUT_NOHB - 5) . " $TEST_WRAPPER '$test' $run_options $inject_code | tee $LOG_DIR/$category/$num; echo \${PIPESTATUS[0]} > $LOG_DIR/subtest_result_num", $TIMEOUT_NOHB);
+        $status_num = script_output("tail -n 1 $LOG_DIR/subtest_result_num");
+        $test_duration = time() - $test_start;
+    };
+    if ($@) {
+        $test_status = 'FAILED';
+        $test_duration = time() - $test_start;
+        sleep(2);
+        copy_log($category, $num, 'out.bad');
+        copy_log($category, $num, 'full');
+        copy_log($category, $num, 'dmesg');
+        copy_fsxops($category, $num);
+        collect_fs_status($category, $num);
+        if (get_var('BTRFS_DUMP', 0) && (check_var 'XFSTESTS', 'btrfs')) { dump_btrfs_img($category, $num); }
+        if (get_var('RAW_DUMP', 0)) { raw_dump($category, $num); }
+
+        prepare_system_shutdown;
+        select_console 'root-console' unless is_pvm;
+        send_key 'alt-sysrq-b';
+        reconnect_mgmt_console if is_pvm;
+        $self->wait_boot;
+
+        sleep 1;
+        select_console('root-console');
+        # Save kdump data to KDUMP_DIR if not set "NO_KDUMP=1"
+        unless (check_var('NO_KDUMP', '1')) {
+            unless (save_kdump($test, $KDUMP_DIR, vmcore => 1, kernel => 1, debug => 1)) {
+                # If no kdump data found, write warning to log
+                my $msg = "Warning: $test crashed SUT but has no kdump data";
+                script_run("echo '$msg' >> $LOG_DIR/$category/$num");
+            }
+        }
+
+        # Add test status to STATUS_LOG file
+        log_add($STATUS_LOG, $test, $test_status, $test_duration);
+
+        # Reload loop device after a reboot
+        if (get_var('XFSTESTS_LOOP_DEVICE')) {
```

Comment:
Also make it maybe oneline statement

Reply:
+1, solved. And also few lines after that about heartbeat_stop, I also changed into one line.